jplag/src/main/java/de/jplag/clustering/algorithm/SpectralClustering.java

```diff
@@ -0,0 +1,172 @@
+
+        // find number of clusters as the multiplicity of eigenvalue 0
+        double eps = 0.05;
+        int minK = Math.max(2, (int) DoubleStream.of(ed.getRealEigenvalues()).filter(x -> x < eps).count());
+        int maxK = (int) Math.ceil(N / 2.0);
+
+        // Find number of clusters using bayesian optimization
+        RealVector lengthScale = new ArrayRealVector(1, options.kernelBandwidth);
+        BayesianOptimization bo = new BayesianOptimization(new ArrayRealVector(1, minK), new ArrayRealVector(1, maxK), options.minRuns,
+                options.maxRuns, options.GPVariance, lengthScale);
+        // bo.debug = true;
+        BayesianOptimization.OptimizationResult<Collection<Collection<Integer>>> boResult = bo.maximize(r -> {
+            int k = (int) Math.round(r.getEntry(0));
+            k = Math.max(minK, k);
+            k = Math.min(maxK, k);
+            Collection<Collection<Integer>> clustering = cluster(k, N, eigenValueIds, ed);
+            IntegerClusteringResult modularityRes = new IntegerClusteringResult(clustering, similarityMatrix);
+            return new BayesianOptimization.OptimizationResult<>(modularityRes.getWorth((a, b) -> (float) similarityMatrix.getEntry(a, b)),
+                    clustering);
+        });
+
+        return boResult.getValue();
+    }
+
+    private Collection<Collection<Integer>> cluster(int k, int n, List<Integer> eigenValueIds, EigenDecomposition ed) {
+        RealMatrix Q = new Array2DRowRealMatrix(n, k);
+        Q.walkInOptimizedOrder(new DefaultRealMatrixChangingVisitor() {
+            public double visit(int row, int column, double value) {
+                int eigenVectorId = eigenValueIds.get(column);
+                RealVector eigenVector = ed.getEigenvector(eigenVectorId);
+                return eigenVector.getEntry(row);
+            };
+        });
+
+        List<ClusterableEigenVector> normRows = IntStream.range(0, n).filter(i -> Q.getRowVector(i).getNorm() > 0)
+                .mapToObj(row -> new ClusterableEigenVector(row, Q.getRowVector(row).unitVector())).collect(Collectors.toList());
+
+        Clusterer<ClusterableEigenVector> clusterer = new KMeansPlusPlusClusterer<>(k, options.maxKMeansIterations);
+        List<? extends Cluster<ClusterableEigenVector>> clusters = clusterer.cluster(normRows);
+        return clusters.stream().map(cluster -> {
+            return cluster.getPoints().stream().map(x -> x.id).collect(Collectors.toList());
+        }).collect(Collectors.toList());
+    }
+
+    private static class ClusterableEigenVector implements Clusterable {
+        private int id;
+        private double[] eigenVector;
+
+        public ClusterableEigenVector(int id, RealVector eigenVector) {
+            this.id = id;
+            this.eigenVector = eigenVector.toArray();
+        }
+
+        @Override
+        public double[] getPoint() {
+            return eigenVector;
+        }
+    }
+
+    public static class ClusteringOptions {
```

Comment:
It seems that different classes in your PR implement their own clustering options class? I am not sure If I understand what the design rationale behind it is. Can you elaborate?

Reply:
Just before reading this comment I deleted the nested ClusteringOptions classes